misc/call_fuzzer.rb

```diff
@@ -0,0 +1,184 @@
+  num_pargs = rand(10)
+  opt_parg_idx = rand(num_pargs)
+  num_opt_pargs = rand( num_pargs + 1 - opt_parg_idx)
+  pargs = []
+  num_pargs.times { |i| pargs.push("p#{i}") }
+  opt_pargs = pargs[opt_parg_idx...opt_parg_idx + num_opt_pargs]
+
+  # Choose how many kwargs to use, and how many are optional
+  num_kwargs = rand(10)
+  num_opt_kwargs = rand(num_kwargs + 1)
+  kwargs = []
+  num_kwargs.times { |i| kwargs.push("k#{i}") }
+  opt_kwargs = kwargs.sample(num_opt_kwargs)
+
+  # Choose whether to have a block argument and splats or not
+  block_arg = rand() < 0.25
+  has_rest = num_opt_pargs == 0 && rand() < 0.5
+  has_kwrest = rand() < 0.25
+
+  #
+  # Generate a method definitions
+  #
+  m_str = "def m("
+
+  pargs.each_with_index do |name, i|
+    if !m_str.end_with?("(")
+      m_str += ", "
+    end
+    m_str += "#{name}"
+
+    # If this has a default value
+    if opt_pargs.include?(name)
+      m_str += " = #{i}"
+    end
+  end
+
+  if has_rest
+    if !m_str.end_with?("(")
+      m_str += ", "
+    end
+    m_str += "*rest"
+  end
+
+  kwargs.each_with_index do |name, i|
+    if !m_str.end_with?("(")
+      m_str += ", "
+    end
+    m_str += "#{name}:"
+
+    # If this has a default value
+    if opt_kwargs.include?(name)
+      m_str += " #{i}"
+    end
+  end
+
+  if has_kwrest
+    if !m_str.end_with?("(")
+      m_str += ", "
+    end
+    m_str += "**kwrest"
+  end
+
+  if block_arg
+    if !m_str.end_with?("(")
+      m_str += ", "
+    end
+    m_str += "&block"
+  end
+
+  m_str += ")\n"
+  if block_arg
+    m_str += "if block; block.call; end\n"
+  end
+  m_str += "if block_given?; yield; end\n"
+
+  if has_rest
+    m_str += "raise 'rest is not array' unless rest.kind_of?(Array)\n"
+    m_str += "raise 'rest size not integer' unless rest.size.kind_of?(Integer)\n"
+  end
+
+  if has_kwrest
+    m_str += "raise 'kwrest is not a hash' unless kwrest.kind_of?(Hash)\n"
+    m_str += "raise 'kwrest size not integer' unless kwrest.size.kind_of?(Integer)\n"
+  end
+
+  m_str += "777\n"
+  m_str += "end"
+
+  #
+  # Generate a random call to the method
+  #
+  c_str = "m("
+
+  pargs.each_with_index do |name, i|
+    if !c_str.end_with?("(")
+      c_str += ", "
+    end
+
+    c_str += "#{i}"
+
+    # TODO: don't always pass optional positional args
+  end
+
+  kwargs.each_with_index do |name, i|
+    # Don't always pass optional kwargs
+    if opt_kwargs.include?(name) && rand() < 0.5
+      next
+    end
+
+    if !c_str.end_with?("(")
+      c_str += ", "
+    end
+
+    c_str += "#{name}: #{i}"
+  end
+
+  c_str += ")"
+
+  # Randomly pass a block or not
+  if rand() < 0.5
+    c_str += " { 1 }"
+  end
+
+  [m_str, c_str]
+end
+
+iseqs_compiled_start = RubyVM::YJIT.runtime_stats[:compiled_iseq_entry]
+
+num_iters.times do |i|
+  puts "Iteration #{i}"
+
+  m_str, c_str = gen_random_method()
+
+  eval("class Foo; end")
+
+  f = Foo.new()
+
+  # Define the method on f
+  puts "Defining"
+  p m_str
+  f.instance_eval(m_str)
```

Comment:
```suggestion
  f = Object.new
  # Define the method on f
  puts "Defining"
  p m_str
  f.instance_eval(m_str)
```

With `instance_eval`, the methods are added to the singleton class, which is already freshly made each time already, so you don't need to make a `Foo` every time. (unnecessary here, but you could also call `Class.new` instead of `eval` to make a new class)

```
$ irb
irb(main):001> o = Object.new
=> #<Object:0x0000000104203a78>
irb(main):002> o.instance_eval('def my_method = 0')
=> :my_method
irb(main):003> o.singleton_class.instance_methods(false)
=> [:my_method]
irb(main):004> Object.instance_methods(false)
=> []
```